core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -70,23 +72,174 @@ object H2OSchemaUtils {
+            fillMethod(mapField, i, Some(qualifiedName))
+          }
+        case ArrayType(elementType, _) =>
+          fillArray(flatSchemaIndexes, buffer, row, elementType, index, qualifiedName)
+        case StructType(fields) =>
+          val subRow = row.getStruct(index)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          fields.zipWithIndex.foreach { case (f, i) => fillMethod(f, i, Some(qualifiedName)) }
+        case dt => buffer(flatSchemaIndexes(qualifiedName)) = row.get(index)
+      }
+    }
+  }
+
+  private def fillArray(
+      flatSchemaIndexes: Map[String, Int],
+      buffer: ArrayBuffer[Any],
+      row: Row,
+      elementType: DataType,
+      index: Int,
+      qualifiedName: String){
+    val seq = row.getSeq[Any](index)
+    val subRow = Row.fromSeq(seq)
+    val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+    (0 until seq.size).foreach{ i =>
+      val arrayField = StructField(i.toString, elementType)
+      fillMethod(arrayField, i, Some(qualifiedName))
+    }
+  }
+
+  def flattenSchema(df: DataFrame): StructType = {
+    val originalSchema = df.schema
+    val fields = df.rdd.map[Seq[FieldWithOrder]] { row: Row =>
+        originalSchema.fields.zipWithIndex.foldLeft(Seq.empty[FieldWithOrder]) {
+          case (acc, (field, index)) => acc ++ flattenField(field, row, index, index :: Nil)
+        }
+      }
+      .reduce { (first, second) =>
+        def convertToMap(collection: Seq[FieldWithOrder]) = collection.map(f => f.order -> f.field).toMap
+        val firstMap = convertToMap(first)
+        val secondMap = convertToMap(second)
+        val pathItemOrdering = new Ordering[Any] {
+          override def compare(x: Any, y: Any): Int = (x, y) match {
+            case (a: Int, b: Int) => a.compareTo(b)
+            case (a: String, b: String) => a.compareTo(b)
+            case (a, b) => a.toString.compareTo(b.toString)
+          }
+        }
+        val keys = (firstMap.keySet ++ secondMap.keySet).toSeq.sorted(Ordering.Iterable(pathItemOrdering))
+        keys.map { key =>
+          (firstMap.get(key), secondMap.get(key)) match {
+            case (None, Some(StructField(name, dataType, _, _))) =>
+              FieldWithOrder(StructField(name, dataType, true), key)
+            case (Some(StructField(name, dataType, _, _)), None) =>
+              FieldWithOrder(StructField(name, dataType, true), key)
+            case (Some(StructField(name, dataType, nullable1, _)), Some(StructField(_, _, nullable2, _))) =>
+              FieldWithOrder(StructField(name, dataType, nullable1 || nullable2), key)
+          }
+        }
+      }
+      StructType(fields.map(_.field))
+  }
+
+  private def getQualifiedName(prefix: Option[String], name: String) = prefix match {
+    case None => name
+    case Some(p) => s"${p}_$name"
+  }
+
+  private def flattenField(
+      originalField: StructField,
+      row: Row,
+      index: Int,
```

Comment:
I would also remove `row` and `index` and pass just row.get(index) for the reasons mentioned above as well